### ISSUE TITLE
Possible to generate random UUID when attaching (inserting row in pivot table)?

### ISSUE THREAD

**Issue by cbj4074** [fico7489/LaravelPivot#35]
Hello!

Thanks for creating this much-needed package!

I don't normally use an Issue to ask a question, but I can't figure out how to do the one thing I had hoped to be able to do with this package.

I have precisely the use-case described in this post:

https://stackoverflow.com/questions/42422068/laravel-eloquent-attach-auto-generate-random-id

In short, I have a master-to-master replication scenario that demands that auto-incremented numeric IDs not be used (due to potential conflicts) on several tables, some of which correspond directly to models, and others of which are pivot tables.

For models, this is solved easily with something like this in the base model class:

```php
/**
 *  Setup model event hooks
 */
public static function boot()
{
    parent::boot();
    
    self::creating(function ($model) {
        $model->id = Uuid::generate();
    });
}
```

But for pivot tables, which don't have a corresponding model, the situation is more complex.

In essence, I need to do something like this whenever I attach the models:

```php
->attach($modelAId, $modelBId, ['id' => Uuid::generate()]);
```

This "works", but I don't want to have to do it explicitly in every instance; it would be far more ideal to use an event-driven approach, as is suggested in the above-cited post; something like this, which is aimed at Laravel 5.4 and no longer seems to work in 5.5:

```php
    public function boot()
    {
    Pivot::creating(function($pivot) {
        if ($pivot->getTable() == 'role_permissions') {
            $pivot->id = Uuid::generate();
        }
    });
```

With this library, I'm certainly able to *access* the value in question, but I can't figure out how to *set* it.

Basically, it seems I would need to do this:

```php
static::pivotAttaching(function ($model, $relationName, $pivotIds, $pivotIdsAttributes) {
    foreach ($pivotIds as $thisRelationName => $thisIdsAttributes) {
        $pivotIdsAttributes[$thisIdsAttributes]['id'] = Uuid::generate();
    }
    
    // But I can't figure out how to "return" $pivotIdsAttributes from here, such that
    // they're used during insertion into the pivot table...
});
```

Am I overlooking something? And if not, is there any way to achieve the desired behavior with this package?

Thanks in advance!

**Comment by fico7489** [comment]
no, you can't set pivot value like you want, when you use sync() laravel does not use Pivot model, it runs query directly, actually this it the reason why this package exists :D

The thing you can do is to foreach all pivot ids and update them, something like this : 

```
    static::pivotAttaching(function ($model, $relationName, $pivotIds, $pivotIdsAttributes) {
        foreach($pivotIds as $id){
             Pivot::find($id)->update(['id' => Uuid::generate()]);
        }
    });
```

**Comment by cbj4074** [comment]
Thanks, @fico7489 ! I really appreciate the reply.

That's an interesting idea! Unfortunately, I don't think it will work for my use-case.

In distributed or replicated database topologies, there should be no possibility for conflicts/collisions among the `id` column values, hence the need to use a UUID in the `id` column for *all three* tables (each of the model tables, and the pivot table).

For the suggested approach to work, the primary key (`id`) column on the pivot table would need to be capable of storing both an auto-incremented integer and a UUID, in order to replace the former with the latter, which isn't possible.

I managed to accomplish what I need without using this package, and it works for `save()` and `attach()`, although I suspect I will encounter further challenges with `sync()`, for the reason you mention.

I'm going to fiddle with this a bit more, and if I manage to get it all working, I'll post a follow-up as to how I accomplished the task at hand. It's difficult to articulate all the nuances and requirements of what I'm working to achieve without source code snippets... maybe once you see some code, you will have other ideas. No obligation, of course! I appreciate your response very much already.

I'll go ahead and close the issue. Thanks again!
